### ISSUE TITLE
Instruct tuning for lora/finetune?

### ISSUE THREAD

**Issue by fblissjr** [ml-explore/mlx-examples#484]
Please correct me if I'm wrong, but it looks like the current examples for lora training all build a loss function around completion, which lines up with the lora example of using only the 'text' field from the jsonl dataset.

Are there any forks or plans to allow for instruct tuning, where the input is an input prompt, and the loss function is targeting the input/output pair?

Or did I miss something?

Thanks!

**Comment by Solido** [comment]
Need confirmation but the input inside text is multiline.
My comprehension is that line return is the input and the completion.

**Comment by fblissjr** [comment]
Looking at the mlx-llm code, I think we need to adjust the Dataset class to account for various dataset types (instruct, chat, etc). Which starts to turn into an axolotl type project eventually, but for simplicity, probably just being able to pass custom dataset types.

Then I think the tuner.trainer needs to be modified for the default loss function (or a new one added for instruct templates).

Let me know if I'm wrong here, but from what I'm seeing, the only training function available right now is for completions.

**Comment by fblissjr** [comment]
I believe this fork handles it correctly: src/mlx_tuning_fork/training.py

**Comment by awni** [comment]
We have a more featured version of lora in mlx-lm https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/LORA.md

Assuming it doesn’t add much code complexity I think it would be cool to update it to support alternative losses / styles of training. Depending on how niche the approach is and/or complex it may also make sense to do it as a standalone package. I’ll mark this issue as enhancement for now.

**Comment by Solido** [comment]
MLX is growing fast and community will soon build around a lot.
Everything that can be common ground for those projects should be welcome.
I'm myself working exclusively on instruct and patiently waiting for more options.

**Comment by chimezie** [comment]
Gentle ping regarding this: #1086
